### PR TITLE
Proposal of alternative mechanism to identify entry nodes

### LINKED ISSUE #43: Relying on aliases is bad
Currently, the specs are telling every client implementation to build a list of entry nodes based on the set of `m.room.aliases` events in a room. This is bogus because:

* as already outlined [here](https://matrix.org/docs/spec/client_server/r0.4.0.html#m-room-aliases), this set isn't bound to reflect the reality of the list of aliases for a room.
* a malicious node could abuse room aliases, by setting an alias to a given room long enough so that it's widely stored in clients, then moving it to another room, which could broadcast propaganda or disinformation.
* there are ongoing discussions within the Matrix core team to get rid of `m.room.aliases` events given they're more painful to manage than actually useful.

One easy solution to work around this issue would be to impose to all client implementations to make sure all room aliases for a given room resolve to the given room.

A more complex but more efficient one would be to stop using room aliases altogether and only rely on room IDs, and build the list of entry nodes based on the list of servers in the room. We could also specify a state event that a node could send in order to indicate whether it wants to be an entry node, but this sounds a bit bogus as well since any user on the node could send that event. Another way of identifying entry nodes is by checking whether each node in the room allows guest registration, but that seems meh to me privacy-wise.

The first solution might be tricky and full of edge cases, and not future-proof, and we don't have a stable release of the specs yet so now is the good time to ditch everything with something completely new.

The only blurry point I can think of regarding the second solution is how to tell the difference between an entry node and a non-entry node. We could specify a state event that a node could send in order to indicate whether it wants to be an entry node, but this sounds a bit bogus since any user on the node could send that event. Another way of identifying entry nodes is by checking whether each node in the room has guest registration turned on, but that seems meh to me privacy-wise.

### PR DESCRIPTION
**Pull request by vabd**
Fixes #43

This proposal introduces a mechanism to discover entry nodes and manage client implementations' lists of entry nodes that doesn't involve using aliases, but instead processing the rest of the room's state and a file in the .well-known namespace.

Rendered: https://specs.informo.network/scs/45/information-distribution/node/

### CODE REVIEW

File: content/information-distribution/node.md [hunk 1]
```diff
@@ -44,81 +44,132 @@ This ping **must** take the form of a Matrix timeline event of the
...
-the Matrix room to join.
-
-Once a Matrix room is joined, client implementations **must** retrieve and save
-the [list of
-aliases](https://matrix.org/docs/spec/client_server/r0.4.0.html#m-room-aliases)
-for this Matrix room. This can be done by processing all of the `m.room.aliases`
-events in the state of the Matrix room, each of them containing the list of
-aliases defined for a given node. The content for such an event looks like this:
+**must** allow guest access and **must** have at least one user in every
+federation they want to give users access to.
```
Comment (GordonF42) [inline content/information-distribution/node.md]: Why is it necessary to have at least one user in the federation?

If there is no users from the server connected to the federation, then the server is simply "unknown" but can still be used as an entry point

File: content/information-distribution/node.md [hunk 2]
```diff
@@ -44,81 +44,132 @@ This ping **must** take the form of a Matrix timeline event of the
...
 
-```
-"aliases": [
-    "#alias1:example.com",
-    "#alias2:example.com"
-]
-```
+On top of that, an entry node **must** expose a file at the address
```
Comment (GordonF42) [inline content/information-distribution/node.md:49]: I don't really like forbidding users to access to Informo through a server that hasn't been explicitly configured for Informo.

File: content/information-distribution/node.md [hunk 3]
```diff
@@ -44,81 +44,132 @@ This ping **must** take the form of a Matrix timeline event of the
...
+users](https://matrix.org/docs/spec/client_server/r0.4.0.html#get-matrix-client-r0-rooms-roomid-joined-members)
+for the federation, then processing this list in order to only keep the server
+part of the Matrix user IDs (i.e. what's after the colon (`:`)). Client
+implementations **must** save this list in case the node they're currently using
+becomes unreachable.
+
+Client implementations **should** weight this list according to the number of
+Matrix users belonging to a given node in a federation. The thinking behind this
+is that a node with many users in a federation is more likely to be an entry
+node for this federation than a node with only a few users in it.
```
Comment (GordonF42) [inline content/information-distribution/node.md]: Not a big fan of this, since servers can create as many fake users as they want, and guests users that are thrown away by clients stay in the room (i.e. an old server will likely have more users than a new popular one).

Also this behaviour encourages silos: if big servers are always put on top, smaller servers won't get as many new users as they should.

File: content/information-distribution/node.md [hunk 4]
```diff
@@ -44,81 +44,132 @@ This ping **must** take the form of a Matrix timeline event of the
...
+
+Client implementations **should** also weight this list accordint to the number
```
Comment (GordonF42) [inline content/information-distribution/node.md]:
```suggestion
Client implementations **should** also weight this list according to the number
```

File: content/information-distribution/node.md [hunk 5]
```diff
@@ -44,81 +44,132 @@ This ping **must** take the form of a Matrix timeline event of the
...
+federation to join.
+
+When reaching a node in order to join a federation, client implementations
+**must** try to retrieve the `/.well-kown/informo/info` file of the node (while
+following 30x redirects). If the file doesn't exist, or doesn't explicitly state
+that the node can be used as an entry node for the given federation, then client
+implementations **must** give up and try another node.
+
+Client implementations **should** store the result of the attempt at retrieving
+the `/.well-kown/informo/info` file of a node in a cache.
```
Comment (GordonF42) [inline content/information-distribution/node.md]: What happens when this file becomes obsolete (changed by the server owner) ? How can we detect it?

File: content/information-distribution/node.md [hunk 6]
```diff
@@ -44,81 +44,132 @@ This ping **must** take the form of a Matrix timeline event of the
...
 action by a user or, either partially or entirely, from a client
 implementation's inner mechanisms, client implementations **must** send a [join
 request](https://matrix.org/docs/spec/client_server/r0.4.0.html#post-matrix-client-r0-join-roomidoralias)
-to the selected node using the selected alias.
+to the selected node using the federation's room ID.
+
+In order to optimise the likelihood of success of the join request, client
+implementations **should** provide at least 3 nodes to join the federation
+through using `server_name` query parameters. Client implementations don't need
+to make sure that these nodes are entry nodes in this case.
```
Comment (GordonF42) [inline content/information-distribution/node.md]: `Client implementations don't need to make sure that these nodes are entry nodes in this case.`

This is a bit ambiguous: you can understand that client implementations provides entry nodes that they are not sure if they are real entry nodes. I'd rephrase it like:
`Those nodes are expected to be entry nodes, and client implementations **can** try to join those nodes without verifying if they are effectively entry nodes.`

**Review by GordonF42 — DISMISSED** [review]

File: content/information-distribution/node.md (hunk 1 above)
Comment (vabd) [reply content/information-distribution/node.md]: True. I wrote it without thinking that a room could be joined through other servers. Will remove this requirement.

File: content/information-distribution/node.md (hunk 2 above)
Comment (vabd) [reply content/information-distribution/node.md:49]: And I really don't like node admins having to handle the legal implications of managing an entry node without their consent. People might set up a HS that's compatible with our entry nodes requirements for their own use case without wishing to take part in Informo, especially if they're operating in a country which current politics aren't keen to freedom of the press.
My point is that, while we need to make sure Informo clients' users live in a perfect world, we also need to look after node admins because Informo is worthless without them.

File: content/information-distribution/node.md (hunk 5 above)
Comment (vabd) [reply content/information-distribution/node.md]: Right, I thought I mentioned that the cache would have to use a time-base invalidation policy, but apparently I didn't.

File: content/information-distribution/node.md (hunk 3 above)
Comment (vabd) [reply content/information-distribution/node.md]: Good point. Not sure what to do here, as we want a way to avoid our list being filled with relay nodes. Maybe set a threshold, e.g. if a node has more than 5 users in the room we should consider it more likely to act as a relay node? wdyt?

File: content/information-distribution/node.md (hunk 6 above)
Comment (vabd) [reply content/information-distribution/node.md]: Thing is, these nodes **aren't** expected to be entry nodes. All we care about is that they are in the room, and the only (indirect) interaction a client can generate with them is if the entry node the client is talking to isn't in the room, then that entry node is going to ask that other node to make it join it. Once it's done the client won't generate any other interaction with this other node, either direct or indirect (unless the other node is an entry node and the client eventually starts using it, or a similar scenario happens).

I should add details in there to make it clearer.

File: content/information-distribution/node.md (hunk 5 above)
Comment (vabd) [reply content/information-distribution/node.md]: Fixed in https://github.com/Informo/specs/pull/45/commits/63c5d7a65f8b10613250a8559ebe919cc7fddea5

File: content/information-distribution/node.md (hunk 1 above)
Comment (vabd) [reply content/information-distribution/node.md]: Fixed in https://github.com/Informo/specs/pull/45/commits/f611b57a8970630f2036565a978ae2165d5be17a

File: content/information-distribution/node.md (hunk 6 above)
Comment (vabd) [reply content/information-distribution/node.md]: Fixed in https://github.com/Informo/specs/pull/45/commits/e070a7792f07b22e58e3426a62573e1cc4baa2bf

File: content/information-distribution/node.md (hunk 3 above)
Comment (GordonF42) [reply content/information-distribution/node.md]: I don't see any good solution either. I'd rather remove this part and let the implementation decide, since it's not really that important anyway (IMO the important thing is ordering by trust counts)

Comment (vabd) [reply content/information-distribution/node.md]: Yeah, that makes sense.

Comment (vabd) [reply content/information-distribution/node.md]: Fixed in 2efd4af

File: content/information-distribution/node.md (hunk 2 above)
Comment (vabd) [reply content/information-distribution/node.md:49]: As per discussion that happened outside of this space, we've come to an agreement on keeping this as is.

File: content/information-distribution/node.md [hunk 7]
```diff
@@ -44,81 +44,129 @@ This ping **must** take the form of a Matrix timeline event of the
...
-aliases](https://matrix.org/docs/spec/client_server/r0.4.0.html#m-room-aliases)
-for this Matrix room. This can be done by processing all of the `m.room.aliases`
-events in the state of the Matrix room, each of them containing the list of
-aliases defined for a given node. The content for such an event looks like this:
+**must** allow guest access.
+
+On top of that, an entry node **must** expose a file at the address
+`https://SERVER_NAME/.well-known/informo/info` which content is a map in the
+JSON format, which associates Matrix room IDs with a map implementing the
+following structure:
```
Comment (GordonF42) [inline content/information-distribution/node.md:52]: I think we should add some kind of blacklist system, so that people can host an Informo entry node for any federation.
Something like:
```json
{
  "federation_list_type": "blacklist",
  "federation_list": {
    "!xxx:example.com": {},
    "!yyy:example.com": {}
  }
}
```

File: content/information-distribution/node.md [hunk 8]
```diff
@@ -44,81 +44,129 @@ This ping **must** take the form of a Matrix timeline event of the
...
+
+| Parameter    | Type   | Req. | Description                                                   |
+|:-------------|:-------|:----:|:--------------------------------------------------------------|
+| `entry_node` | `bool` |  x   | Whether the node is an entry node for this federation (room). |
```
Comment (GordonF42) [inline content/information-distribution/node.md:56]: I think the `entry_node` property should be in the root of the structure, since this is server configuration related, and not room-related.
What's your opinion?

File: content/information-distribution/node.md [hunk 9]
```diff
@@ -44,81 +44,129 @@ This ping **must** take the form of a Matrix timeline event of the
...
+
+Client implementations **must** provide users with the room ID of at least one
+federation, and allow users to input custom values for the room ID of the
+federation to join.
+
+When reaching a node in order to join a federation, client implementations
+**must** try to retrieve the `/.well-kown/informo/info` file of the node (while
+following 30x redirects). If the file doesn't exist, or doesn't explicitly state
+that the node can be used as an entry node for the given federation, then client
+implementations **must** give up and try another node.
```
Comment (GordonF42) [inline content/information-distribution/node.md]: Maybe add a note saying this is not a technical requirement, but rather an ethical requirement to prevent servers from joining unwanted informo federations, which could cause legal issues and force the server to be closed.

File: content/information-distribution/node.md [hunk 10]
```diff
@@ -44,81 +44,129 @@ This ping **must** take the form of a Matrix timeline event of the
...
+#### Reaching an entry node
+
+Because a list of entry nodes only contains the nodes' server names, and not
+their host names, a node might not be reachable at the address defined by its
 server name. As an example, the Matrix specifications would allow a node living
-at `node.example.com` to register the alias `#example:example.com`.
+at `node.example.com` to use Matrix identifiers ending with `:example.com`.
 
-In order to find the effective FQDN and port to reach a node at, client
+In order to find the effective host name and port to reach a node at, client
```
Comment (GordonF42) [inline content/information-distribution/node.md]: Maybe replace `host name` to `address` ?

File: content/information-distribution/node.md [hunk 11]
```diff
@@ -44,81 +44,129 @@ This ping **must** take the form of a Matrix timeline event of the
...
 order to join a federation, whether this selection originated from an explicit
 action by a user or, either partially or entirely, from a client
 implementation's inner mechanisms, client implementations **must** send a [join
 request](https://matrix.org/docs/spec/client_server/r0.4.0.html#post-matrix-client-r0-join-roomidoralias)
-to the selected node using the selected alias.
+to the selected node using the federation's room ID.
+
+In order to optimise the likelihood of success of the join request, client
+implementations **should** provide server names for at least additional 3 nodes
+when performing the join request using `server_name` query parameters. This will
```
Comment (GordonF42) [inline content/information-distribution/node.md:191]: ` implementations **should** provide server names for at least additional 3 nodes when performing the join request using `server_name` query parameters`

I don't really get what are the "server_name query parameters".

I would expect the client implementation to provide 3 or more informo entrypoints, and let the user chose which server to join, but that's not what I understand with this phrase.

**Review by GordonF42 — DISMISSED** [review]

File: content/information-distribution/node.md (hunk 8 above)
Comment (vabd) [reply content/information-distribution/node.md:56]: My thinking behind this is that a node can want to be an entry node for some federation but not all of them (e.g. "I want to be an entry node for the `informo.network` federation but not that neo-nazi one right there).

File: content/information-distribution/node.md (hunk 11 above)
Comment (vabd) [reply content/information-distribution/node.md:191]: Link is above. As discussed in private chat, this is clear enough.

File: content/information-distribution/node.md (hunk 7 above)
Comment (vabd) [reply content/information-distribution/node.md:52]: Setting `entry_node` to `false` for a federation already works as a way to blacklist federations.

File: content/information-distribution/node.md (hunk 9 above)
Comment (vabd) [reply content/information-distribution/node.md]: Good point.

File: content/information-distribution/node.md (hunk 7 above)
Comment (vabd) [reply content/information-distribution/node.md:52]: Conclusions from private chat: we'll need to couple that with a `*` wildcard value for the federation which would allow one to define either a blacklist of a whitelist of entry nodes with the same content format.

File: content/information-distribution/node.md (hunk 10 above)
Comment (vabd) [reply content/information-distribution/node.md]: Yes

File: content/information-distribution/node.md [hunk 12]
```diff
@@ -44,81 +44,158 @@ This ping **must** take the form of a Matrix timeline event of the
...
+On top of that, an entry node **must** expose a file at the address
+`https://SERVER_NAME/.well-known/informo/info` which content is a map in the
+JSON format, which associates Matrix room IDs with a map implementing the
+following structure:
+
+| Parameter    | Type   | Req. | Description                                                   |
+|:-------------|:-------|:----:|:--------------------------------------------------------------|
+| `entry_node` | `bool` |  x   | Whether the node is an entry node for this federation (room). |
+
+One entry of this map **can** be the wildcard value "`*`" instead of a Matrix
```
Comment (GordonF42) [inline content/information-distribution/node.md:58]: Another way to see it is that the `*` object is the default value for informo federation objects, and the other objects overrides this default value on a per-field basis. 

The current sentense is OK for me, but you can change it if you think it can be better formulated.

**Review by GordonF42 — DISMISSED** [review]
LGTM :+1: 
You can reword the sentence on the comment if you deem it necessary, or merge this PR in its current state

**Review by GordonF42 — APPROVED** [review]
